### PR TITLE
[GEOS-11543] WFS Unable to use propertyName to filter properties in a GetFeature request

### PR DESCRIPTION
**Pull request by turingtestfail**
[GEOS-11543](https://osgeo-org.atlassian.net/browse/GEOS-11543) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)

fixed to ignore lists of lists and added tests

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Comment by jratike80** [comment]
Do you know when this issue has come? We have even an example in the standard demo requests "WFS_getFeatureBetweenCQL.url"
http://localhost:8080/geoserver/wfs?request=GetFeature&typeName=topp:states&propertyName=STATE_NAME,LAND_KM,the_geom&version=1.0.0

Maybe there could be also a JIRA issue that explains which versions are/were affected. I feel that &propertyName used to work.

**Comment by turingtestfail** [comment]
> Do you know when this issue has come? We have even an example in the standard demo requests "WFS_getFeatureBetweenCQL.url" http://localhost:8080/geoserver/wfs?request=GetFeature&typeName=topp:states&propertyName=STATE_NAME,LAND_KM,the_geom&version=1.0.0
> 
> Maybe there could be also a JIRA issue that explains which versions are/were affected. I feel that &propertyName used to work.

Not sure the history of this but the issue is not directly with propertyName.  propertyName works but the "service" KVP has to be set, otherwise the parser for propertyName is not found.  The fix I just added to this PR will insert the service KVP if servlet path starts with "wfs/" and we are not in strict CITE mode.

File: src/ows/src/main/java/org/geoserver/ows/Dispatcher.java [hunk 1]
```diff
@@ -1530,6 +1530,14 @@ void parseKVP(Request req) throws ServiceException {
     }
 
     Map<String, Object> parseKVP(Request req, Map<String, Object> kvp) {
+        // if the service is not explicitly set, some WFS parsers will not be found
+        if (!kvp.containsKey("service")
+                && !kvp.containsKey("SERVICE")
```
Comment (aaime) [inline src/ows/src/main/java/org/geoserver/ows/Dispatcher.java:1535]: The KVP map shoud be a CaseInsensitivemap. Little known fact, all keys in OGC services are case insensitive, one can write a valid request with "sErViCe=WFS". However, values are case sensitive, so "service=wfs" is actually not valid, strictly speaking)

File: src/ows/src/main/java/org/geoserver/ows/Dispatcher.java [hunk 2]
```diff
@@ -1530,6 +1530,14 @@ void parseKVP(Request req) throws ServiceException {
...
+                && req.getHttpRequest().getServletPath().toLowerCase().startsWith("/wfs")
```
Comment (aaime) [inline src/ows/src/main/java/org/geoserver/ows/Dispatcher.java:1536]: Good start but I'd generalize it a bit, to make it valid across all OGC services.

See what the Dispatcher is already doing [here](https://github.com/geoserver/geoserver/blob/ce22d0d99c89bc265c64e61dda1340ce1ebf63ef/src/ows/src/main/java/org/geoserver/ows/Dispatcher.java#L587) and [here].(https://github.com/geoserver/geoserver/blob/main/src/ows/src/main/java/org/geoserver/ows/Dispatcher.java#L1597). 
It's basically assuming the request path is the service, it if could not be determined in any other way. 
I would do the same, and ignore the path if it happens to be the generic "ows".

This way you have a fix that generalizes for wms, wfs, wcs and so on (take the path, make it uppercase, use as service).

**Review by aaime — CHANGES_REQUESTED** [review]

File: src/ows/src/main/java/org/geoserver/ows/Dispatcher.java (hunk 2 above)
Comment (turingtestfail) [reply src/ows/src/main/java/org/geoserver/ows/Dispatcher.java]: Ok, I piggybacked off of the method you pointed me too and expanded it a bit with a check on the HttpRequest if the path is not set (which was the case with KVP parse)

**Comment by jratike80** [comment]
Ok, I did not notice that this happens only when the GetFeature request is missing `service=WFS`. It is wrong by the WFS standard because "service" has always been mandatory, but unfortunately GeoServer has always been lenient with the missing parameter.

**Pull request merged**